Improve ability pattern compilation for new runtime

dolio opened the pull request:
Previously the ability pattern compilation was treating it almost identically to data types, where you match on exactly one ability signature in a pattern column. However, you can actually match on 0 or more abilities simultaneously. This fixes that issue in the pattern compiler, and tweaks subsequent phases to handle things properly. Should fix issue #1663.

dolio committed:
Rework ability matching compilation

- Previous pattern matching code was assuming that ability matches
  target exactly one ability. However, they may actually cover any
  number (0 or more). So, instead of determining and emitting code
  for a particular ability, we figure out all the abilities in play.
- This also involves some different handling for trivial, pure-only
  matches. We emit different code that doesn't actually do any
  ability management for these cases.

dolio committed:
Take references into account when compiling patterns

- Tag alone is not sufficient to split a pattern matrix on a request.
  We must make sure the request is from the same ability.

pchiusano commented:
@dolio can you just add a regression test for this? If we're not already doing this, we should be running the tests with both the old and the new runtime. 

Other than that looks good.

dolio committed:
Add to abilities.u test case

- Multiple abilities handled at once, as well as no abilities
- New runtime fails one of the cases. Looking into it.

dolio committed:
Fix pure-only matching in new runtime

- This case was getting rewritten into a non-matching function, but the
  argument that would be the scrutinee is delayed by the handler call.
  So it needs to be forced or we can end up with erroneously delayed
  values leaking into situations where they don't make sense.

dolio committed:
More substantial AccumPure handling based on AccumRequest

- The request examines the term being scrutinized to avoid cases like:
    v = tm
    u = v
    ...
  So it seems reasonable for the pure-only case to do so as well.

dolio commented:
I beefed up the ability test case to check for stuff related to this fix, and it actually works on the new runtime now.

pchiusano commented:
CI isn't currently running the tests with new runtime but #1673 is tracking this. @dolio has manually run the test with new runtime locally and it worked fine.
